PX4/failsafe-overhaul.md

```diff
@@ -0,0 +1,136 @@
+- Feature Name: failsafe_overhaul
+- Start Date: 2022-07-19
+- RFC PR: [dronecode/rfc#0000](https://github.com/dronecode/rfc/pull/0000)
+
```

Comment:
This PR should be added as a reference: https://github.com/PX4/PX4-Autopilot/issues/13719

Yes, failsafes is a good area for exploration. To consider (does not feel addressed):
- Failsafe recovery - what is recoverable and what is not. What is the recovery behaviour.
- Failsafe priority and daisy-chaining. It is possible for multiple failsafes to trigger. How are these handled.

Reply:
@hamishwillee Thanks for the considerations.

I hadn't thought about recovery behavior. I'll think about it and update the RFC when I have some ideas.
About priority:
Yes, it is possible for multiple failsafes to trigger. This is necessary to handle unrelated failsafes that may trigger due to orthogonal failures. Originally, I had thought prioritizing failsafes would solve any problems caused here. However, this does bring up some questions that aren't addressed by my current logic:
* If failsafes are strictly prioritized, how does an orthogonal (lower priority) failsafe interact with an already engaged higher priority failsafe? Ideally, actions like notifying the user and companion computer should still go through, but we don't want a lower priority mode-change overriding a higher priority one.
* How do failsafes of the same priority work?

Reply:
You can see a matrix of behaviour here: https://github.com/PX4/PX4-Autopilot/issues/13719

I'm not sure of solutions. Just raising the issue :-)